File: C/matlabEngineCaller_customLayers.m
function [output,sRough] = matlabEngineCaller_customLayers(params,contrast,funcName,funcPath,bulkIn,bulkOut)

    % ____________________ REMOVE GHOST CODE AT SOME POINT ____________________

paramsLen = int16(length(params));
% dotM = strfind(funcName,'.m');
% if ~isempty(dotM)
%     funcName = funcName(1:dotM-1);
% end
%[path,funcName,extension] = fileparts(funcName);
funName = ['[total,layers] = ' funcName '(params,bulk_in,bulk_out,contrast)'];
if ~isempty(params) 
    sRough= params(1);
end

pathCall= ['cd ''' funcPath ''';'];
% add FuncPath to the path using addPath()

% if funcPath is 'matlabroot', dont add it to path since this is only used for start/stop of matlab engine
if ~strcmp(funcPath,'matlabroot')
    addpath(funcPath);
end


debug = 0;
if coder.target('MATLAB')
    if debug == 1
        disp('Debug');
        disp(funName);
        disp(pathCall);
        layers= 1;
    end
    dotM = strfind(funcName,'.m');
    if ~isempty(dotM)
        funcName = funcName(1:dotM-1);
    end
    fileHandle = str2func(funcName);
    [output,layers] = fileHandle(params,bulkIn,bulkOut,contrast);
else
    
   % p = mfilename('fullpath');
    %extension = '.m';
    %path = strcat(p,extension);


    %$path = pwd;
    %incPath1 = fullfile(matlabroot,'extern','include');

    % ------------------------PATH SETTING FILE SETTING FOR CODER-------------------------------------- 
    % matlabRoot = string(matlabroot) FOR SOME REASON THIS DIDNT WORK IN CODER> ERROR: Cannot pass a mxArray to 'string'.
    %matlabRoot = string(feval('matlabroot')) ;
    %matlabRoot = $matlabroot
    
    

    %m = matlabroot;
    %matlabRoot = char(m);
    source1 = 'matlabCallFun.c';
    
    if ismac % MacOs
        arch = string(computer('arch'));
        incPath1= matlabRoot+'/'+"extern"+'/'+string(arch)+'/'+"include";
        %incPath1 = append(matlabroot,'/','extern','/','include');

        linkPath1 = matlabRoot+'/'+arch+'/'+"glnxa64";

        %linkPath1 = append(matlabroot,'/',arch,'/','glnxa64')    APPEND,FULLFILE and STRCAT are not supported in MATLAB CODER
        %fullfile(matlabroot,arch,'glnxa64'); % NEED TO CONFIRM THIS
        linkFile1 = 'libeng.dylib';
        linkFile2 = 'libmx.dylib';
        
    elseif isunix %LINUX
        arch = string(computer('arch'));
        incPath1= matlabRoot+'/'+"extern"+'/'+"include";

        %incPath1 = append(matlabroot,'/','extern','/','include');

        linkPath1 = matlabRoot+'/'+arch+'/'+"glnxa64";
        %linkPath1 = append(matlabroot,'/',arch,'/','glnxa64');
        %fullfile(matlabroot,arch,'glnxa64');
        linkFile1 = 'libeng.so';
        linkFile2 = 'libmx.so';
        
    elseif ispc % WINDOWS
        %get the required paths for include and link files from a json file called 'customPaths.json' in
        % the RAT main directory

        %open the json file
        %fid = fopen('customPaths.json');
        %read the json file
        %txt = fread(fid,inf);
        %close the file
        %fclose(fid);

        % get the key



        arch = string(computer('arch'));
        %incPath1= matlabRoot+'\'+"extern"+'\'+"include";
        incPath1= 'C:/Program Files/MATLAB/R2021a/extern/include'

        linkPath1 = 'C:/Program Files/MATLAB/R2021a/bin/win64';
        %linkPath1 = fullfile(matlabroot,'extern','include','lib',arch,'mingw64');
        linkFile1 = 'libeng.dll';
        linkFile2 = 'libmx.dll';
    else
        arch = 'undefined';
        incPath1 = 'undefined';
        linkPath1 = 'undefined';
        linkFile1 = 'thisWontWork.lib';
        linkFile2 = 'thisWontWork.lib';
        % DEFINE SOMETHING TO MAKE COMPILER HAPPY
    end
    

    
    %path = 'C:\Users\oba7931403\source\repos\RAT-Sandbox\Sandbox\matlabEngineCaller\tmwExampleNew';
    %incPath1 = fullfile(matlabroot,'extern','include')
    
    %incPath1 = 'C:\Program Files\MATLAB\R2021a\extern\include';
    %incPath2 = '/usr/include/openmpi-x86_64';
    %linkPath1 = fullfile(matlabroot,'extern','include','lib','win64','mingw64')
    %linkPath1 = 'C:\Program Files\MATLAB\R2021a\extern\lib\win64\mingw64';
    %linkFile1 = 'libeng.lib';
    %linkFile2 = 'libmx.lib';
    
    
    %source2 = 'matlabCallFun.h';
    
    %source1 = 'matlabCallFun.c';
    %source2 = 'matlabCallFun.h';
    
    libPriority = '';
    libPreCompiled = true;
    libLinkOnly = true;

    %coder.cinclude(source2);
    coder.updateBuildInfo('addSourceFiles',source1);
    %coder.updateBuildInfo('addSourceFiles',source2);
    coder.updateBuildInfo('addSourcePaths','C:\Users\oba7931403\Documents\GitHub\RAT\C');
    coder.updateBuildInfo('addSourcePaths','$(START_DIR)\C');
    coder.updateBuildInfo('addIncludePaths',incPath1);
    %coder.updateBuildInfo('addIncludePaths',incPath2);
    coder.updateBuildInfo('addLinkObjects',linkFile1,linkPath1,libPriority,libPreCompiled,libLinkOnly);
    coder.updateBuildInfo('addLinkObjects',linkFile2,linkPath1,libPriority,libPreCompiled,libLinkOnly);
    
    %Need to reserve some meory for the referencenced variables
    outp = zeros(1000,3);
    nLayers = 0;
    
    coder.ceval('matlabCallFun', params, paramsLen, funName, pathCall, bulkIn, bulkOut, contrast, coder.wref(outp), coder.wref(nLayers));
    
    rowCount = 1;
    colCount = 1;
    
    m = size(outp,1);
    n = size(outp,3);
    
    %reshape the output to [layers * 3] array
    if (nLayers >= 1)
        output = zeros(nLayers,3);
        for i = 1:(nLayers*3)
            thisVal = outp(i);  %Make use of Matlab linear indexing.
            output(rowCount,colCount) = thisVal;
            rowCount = rowCount + 1;
            if rowCount == (nLayers+1)
                rowCount = 1;
                colCount = colCount + 1;
            end
        end
    else
        output = [1 1 1];
        sRough = 1;
    end

end

File: dev/matlabEngineCaller/codegen/lib/matlabEngineCaller_customLayers/matlabEngineCaller_customLayers.c
/*
 * Non-Degree Granting Education License -- for use at non-degree
 * granting, nonprofit, educational organizations only. Not for
 * government, commercial, or other organizational use.
 *
 * matlabEngineCaller_customLayers.c
 *
 * Code generation for function 'matlabEngineCaller_customLayers'
 *
 */

/* Include files */
#include "matlabEngineCaller_customLayers.h"
#include "matlabEngineCaller_customLayers_emxutil.h"
#include <string.h>

/* Function Definitions */
void matlabEngineCaller_customLayers(const double params[8], double contrast,
  const char funcName[10], const char funcPath[54], double bulkIn, double
  bulkOut, emxArray_real_T *output, double *sRough)
{
  int i;
  char cv[61];
  static const char cv1[17] = { '[', 't', 'o', 't', 'a', 'l', ',', 'l', 'a', 'y',
    'e', 'r', 's', ']', ' ', '=', ' ' };

  char cv2[60];
  static const char cv3[34] = { '(', 'p', 'a', 'r', 'a', 'm', 's', ',', 'b', 'u',
    'l', 'k', '_', 'i', 'n', ',', 'b', 'u', 'l', 'k', '_', 'o', 'u', 't', ',',
    'c', 'o', 'n', 't', 'r', 'a', 's', 't', ')' };

  double b_params[8];
  double outp[300];
  double nLayers;
  unsigned int rowCount;
  int colCount;
  int loop_ub;

  /*  dotM = strfind(funcName,'.m'); */
  /*  if ~isempty(dotM) */
  /*      funcName = funcName(1:dotM-1); */
  /*  end */
  /* [path,funcName,extension] = fileparts(funcName); */
  *sRough = params[0];

  /* incPath2 = '/usr/include/openmpi-x86_64'; */
  /* source2 = '/home/arwel/Documents/RascalDev/RAT/targetFunctions/common/matlabEngineCaller_customLayers/matlabCallFun.h'; */
  /* source1 = 'matlabCallFun.c'; */
  /* source2 = 'matlabCallFun.h'; */
  /* coder.cinclude(source2); */
  /* coder.updateBuildInfo('addSourceFiles',source2); */
  /* coder.updateBuildInfo('addIncludePaths',incPath2); */
  /* Need to reserve some meory for the referencenced variables */
  for (i = 0; i < 17; i++) {
    cv[i] = cv1[i];
  }

  for (i = 0; i < 10; i++) {
    cv[i + 17] = funcName[i];
  }

  for (i = 0; i < 34; i++) {
    cv[i + 27] = cv3[i];
  }

  cv2[0] = 'c';
  cv2[1] = 'd';
  cv2[2] = ' ';
  cv2[3] = '\'';
  for (i = 0; i < 54; i++) {
    cv2[i + 4] = funcPath[i];
  }

  cv2[58] = '\'';
  cv2[59] = ';';
  memcpy(&b_params[0], &params[0], 8U * sizeof(double));
  matlabCallFun(b_params, 8, cv, cv2, bulkIn, bulkOut, contrast, outp, &nLayers);
  rowCount = 1U;
  colCount = 0;

  /* reshape the output to [layers * 3] array */
  if (nLayers >= 1.0) {
    i = (int)nLayers;
    loop_ub = output->size[0] * output->size[1];
    output->size[0] = i;
    output->size[1] = 3;
    emxEnsureCapacity_real_T(output, loop_ub);
    loop_ub = i * 3;
    for (i = 0; i < loop_ub; i++) {
      output->data[i] = 0.0;
    }

    i = (int)(nLayers * 3.0);
    for (loop_ub = 0; loop_ub < i; loop_ub++) {
      output->data[((int)rowCount + output->size[0] * colCount) - 1] =
        outp[loop_ub];

      /* Make use of Matlab linear indexing. */
      rowCount++;
      if (rowCount == nLayers + 1.0) {
        rowCount = 1U;
        colCount++;
      }
    }
  } else {
    i = output->size[0] * output->size[1];
    output->size[0] = 1;
    output->size[1] = 3;
    emxEnsureCapacity_real_T(output, i);
    output->data[0] = 1.0;
    output->data[1] = 1.0;
    output->data[2] = 1.0;
    *sRough = 1.0;
  }
}

/* End of code generation (matlabEngineCaller_customLayers.c) */
